README.md
[![Greenkeeper badge](https://badges.greenkeeper.io/tmcw/relative-luminance.svg)](https://greenkeeper.io/)
[![CircleCI](https://circleci.com/gh/tmcw/relative-luminance/tree/master.svg?style=svg)](https://circleci.com/gh/tmcw/relative-luminance/tree/master)

## relative-luminance

    npm install relative-luminance

Calculate the [relative luminance](http://www.w3.org/TR/2008/REC-WCAG20-20081211/#relativeluminancedef) of
an RGB triplet color.

Note that as of 1.0.0, this module doesn't directly export a function if you're using
CommonJS: use the `default` export, like:

```js
var luminance = require('relative-luminance').default;
```

### API

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

##### Table of Contents

-   [relativeLuminance](#relativeluminance)
    -   [Parameters](#parameters)
    -   [Examples](#examples)

#### relativeLuminance

Given a 3-element array of R, G, B varying from 0 to 255, return the luminance
as a number from 0 to 1.

##### Parameters

-   `rgb` **[Array](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)>** 3-element array of a color

##### Examples

```javascript
var luminance = require('relative-luminance');
var black_lum = luminance([0, 0, 0]); // 0
```

Returns **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** luminance, between 0 and 1


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="2.0.1"></a>
## [2.0.1](https://github.com/tmcw/relative-luminance/compare/v2.0.0...v2.0.1) (2018-12-30)



<a name="2.0.0"></a>
# [2.0.0](https://github.com/tmcw/relative-luminance/compare/v1.0.2...v2.0.0) (2018-08-20)


### Features

* UMD builds, simplified stack ([0bef765](https://github.com/tmcw/relative-luminance/commit/0bef765))


### BREAKING CHANGES

* completely new main, unpkg, and module entry points.



<a name="1.0.2"></a>
## [1.0.2](https://github.com/tmcw/relative-luminance/compare/v1.0.1...v1.0.2) (2018-04-16)


### Bug Fixes

* move esm dependency out devDependencies ([6d6d5bf](https://github.com/tmcw/relative-luminance/commit/6d6d5bf))



<a name="1.0.1"></a>
## [1.0.1](https://github.com/tmcw/relative-luminance/compare/v1.0.0...v1.0.1) (2018-03-21)


### Bug Fixes

* Be conservative about files included in the package. ([f96e583](https://github.com/tmcw/relative-luminance/commit/f96e583))



<a name="1.0.0"></a>
# [1.0.0](https://github.com/tmcw/relative-luminance/compare/v0.0.2...v1.0.0) (2018-03-21)


### Features

* JS modules: use esm to provide both CommonJS and ES6 entry points ([2984a5f](https://github.com/tmcw/relative-luminance/commit/2984a5f))



<a name="0.0.2"></a>
## [0.0.2](https://github.com/tmcw/relative-luminance/compare/v0.0.1...v0.0.2) (2017-05-03)



<a name="0.0.1"></a>
## 0.0.1 (2017-05-03)


test/test.js
const test = require("tape");
const luminance = require("../");

test("luminance", t => {
  t.equal(luminance([0, 0, 0]), 0, "black");
  t.equal(luminance([255, 255, 255]), 1, "white");
  t.end();
});


test/luminance.test.js
import expect from "expect.js";
import luminance from "../index.mjs";
import jsc from "jsverify";

describe("luminance", function() {
  it("calculates lum of black", function() {
    expect(luminance([0, 0, 0])).to.eql(0);
  });
  it("calculates lum of white", function() {
    expect(luminance([255, 255, 255])).to.eql(1);
  });
  jsc.property(
    "always is between 0 and 1",
    jsc.tuple([jsc.integer(0, 255), jsc.integer(0, 255), jsc.integer(0, 255)]),
    function(rgb) {
      return luminance(rgb) >= 0 && luminance(rgb) <= 1;
    }
  );
});


index.mjs
// # Relative luminance
//
// http://www.w3.org/TR/2008/REC-WCAG20-20081211/#relativeluminancedef
// https://en.wikipedia.org/wiki/Luminance_(relative)
// https://en.wikipedia.org/wiki/Luminosity_function
// https://en.wikipedia.org/wiki/Rec._709#Luma_coefficients

// red, green, and blue coefficients
var rc = 0.2126,
  gc = 0.7152,
  bc = 0.0722,
  // low-gamma adjust coefficient
  lowc = 1 / 12.92;

function adjustGamma(_) {
  return Math.pow((_ + 0.055) / 1.055, 2.4);
}

/**
 * Given a 3-element array of R, G, B varying from 0 to 255, return the luminance
 * as a number from 0 to 1.
 * @param {Array<number>} rgb 3-element array of a color
 * @returns {number} luminance, between 0 and 1
 * @example
 * var luminance = require('relative-luminance');
 * var black_lum = luminance([0, 0, 0]); // 0
 */
export default function relativeLuminance(rgb) {
  var rsrgb = rgb[0] / 255;
  var gsrgb = rgb[1] / 255;
  var bsrgb = rgb[2] / 255;

  var r = rsrgb <= 0.03928 ? rsrgb * lowc : adjustGamma(rsrgb),
    g = gsrgb <= 0.03928 ? gsrgb * lowc : adjustGamma(gsrgb),
    b = bsrgb <= 0.03928 ? bsrgb * lowc : adjustGamma(bsrgb);

  return r * rc + g * gc + b * bc;
}


index.js
// # Relative luminance
//
// http://www.w3.org/TR/2008/REC-WCAG20-20081211/#relativeluminancedef
// https://en.wikipedia.org/wiki/Luminance_(relative)
// https://en.wikipedia.org/wiki/Luminosity_function
// https://en.wikipedia.org/wiki/Rec._709#Luma_coefficients

// red, green, and blue coefficients
const rc = 0.2126;
const gc = 0.7152;
const bc = 0.0722;
// low-gamma adjust coefficient
const lowc = 1 / 12.92;

function adjustGamma(_) {
  return Math.pow((_ + 0.055) / 1.055, 2.4);
}

/**
 * Given a 3-element array of R, G, B varying from 0 to 255, return the luminance
 * as a number from 0 to 1.
 * @param {Array<number>} rgb 3-element array of a color
 * @returns {number} luminance, between 0 and 1
 * @example
 * var luminance = require('relative-luminance');
 * var black_lum = luminance([0, 0, 0]); // 0
 */
export default function relativeLuminance(rgb) {
  const rsrgb = rgb[0] / 255;
  const gsrgb = rgb[1] / 255;
  const bsrgb = rgb[2] / 255;

  const r = rsrgb <= 0.03928 ? rsrgb * lowc : adjustGamma(rsrgb);
  const g = gsrgb <= 0.03928 ? gsrgb * lowc : adjustGamma(gsrgb);
  const b = bsrgb <= 0.03928 ? bsrgb * lowc : adjustGamma(bsrgb);

  return r * rc + g * gc + b * bc;
}
